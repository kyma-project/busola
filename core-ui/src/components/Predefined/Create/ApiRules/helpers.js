import * as jp from 'jsonpath';

// hostname may contain optional namespace prefix ({namespace}/{hostname})
function removeNamespacePrefix(host) {
  return host.includes('/') ? host.split('/')[1] : host;
}

export function hasWildcard(host) {
  if (!host) return false;

  // namespace prefix may be also '*' (all namespaces), get rid of it
  return removeNamespacePrefix(host).includes('*');
}

// find gateway by name in format {namespace}/{name}
export function findGateway(gatewayStr, allGateways) {
  gatewayStr = gatewayStr.replace('.svc.cluster.local', '');
  const [gatewayName, gatewayNamespace] = (gatewayStr || '.').split('.');

  return (allGateways || []).find(
    ({ metadata }) =>
      metadata.name === gatewayName && metadata.namespace === gatewayNamespace,
  );
}

export function getGatewayHosts(gateway) {
  const filterUnique = (e, i, arr) => arr.indexOf(e) === i;

  return (gateway?.spec.servers.flatMap(s => s.hosts) || [])
    .map(removeNamespacePrefix)
    .filter(filterUnique);
}

function validateAccessStrategy(accessStrategy) {
  if (accessStrategy?.handler === 'jwt') {
    return (
      accessStrategy?.jwks_urls?.length > 0 &&
      accessStrategy?.trusted_issuers?.length > 0
    );
  } else if (accessStrategy?.handler === 'oauth2_introspection') {
    return accessStrategy?.required_scope?.length > 0;
  } else {
    return true;
  }
}

function validateRule(rule) {
  const areMethodsValid = rule?.methods?.length > 0;

  const accessStrategies = rule?.accessStrategies;
  const areAccessStrategiesValid =
    accessStrategies && accessStrategies.every(validateAccessStrategy);

  return areMethodsValid && areAccessStrategiesValid;
}

export function validateApiRule(apiRule) {
  const hasValidService =
    jp.value(apiRule, '$.spec.service.name') &&
    jp.value(apiRule, '$.spec.service.port');

  const rules = jp.value(apiRule, '$.spec.rules');
  const hasValidRules = rules && rules.every(validateRule);

  return hasValidService && hasValidRules;
}
