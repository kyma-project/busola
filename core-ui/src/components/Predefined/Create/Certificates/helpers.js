import * as jp from 'jsonpath';

export function toYaml(cert) {
  const csrData = cert.withCSR
    ? {
        csr: cert.csr || undefined,
      }
    : {
        commonName: cert.commonName || undefined,
        dnsNames: cert.dnsNames.length ? cert.dnsNames : undefined,
      };

  return {
    apiVersion: 'cert.gardener.cloud/v1alpha1',
    kind: 'Certificate',
    metadata: {
      name: cert.name,
      namespace: cert.namespace,
    },
    spec: {
      ...csrData,
      issuerRef: cert.issuerRef.name ? cert.issuerRef : undefined,
      renew: cert.renew || undefined,
      secreName:
        !cert.existingSecret && cert.secretName ? cert.secretName : undefined,
      secretRef:
        cert.existingSecret && cert.secretRef.name ? cert.secretRef : undefined,
    },
  };
}

export function fromYaml(yaml, prevIssuer) {
  return {
    name: jp.value(yaml, '$.metadata.name') || '',
    namespace: jp.value(yaml, '$.metadata.name') || '',
    commonName: jp.value(yaml, '$.spec.commonName') || '',
    csr: jp.value(yaml, '$.spec.csr') || '',
    dnsNames: jp.value(yaml, '$.spec.dnsNames') || [],
    issuerRef: {
      name: jp.value(yaml, '$.spec.issuerRef.name') || '',
      namespace: jp.value(yaml, '$.spec.issuerRef.namespace') || '',
    },
    renew: jp.value(yaml, '$.spec.renew') || false,
    existingSecret: !!jp.value(yaml, '$.spec.secretRef'),
    secretName: jp.value(yaml, '$.spec.secretName') || '',
    secretRef: {
      name: jp.value(yaml, '$.spec.secretRef.name') || '',
      namespace: jp.value(yaml, '$.spec.secretRef.namespace') || '',
    },
  };
}

export function createTemplate(namespace) {
  return {
    withCSR: true,
    name: '',
    namespace,
    commonName: '',
    csr: '',
    dnsNames: [],
    issuerRef: {
      name: '',
      namespace: '',
    },
    renew: false,
    existingSecret: false,
    secretName: '',
    secretRef: {
      name: '',
      namespace: '',
    },
  };
}
