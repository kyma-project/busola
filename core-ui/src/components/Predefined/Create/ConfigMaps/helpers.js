export function createConfigMapTemplate(namespace) {
  return {
    apiVersion: 'v1',
    kind: 'ConfigMap',
    metadata: {
      namespace,
    },
    data: {},
  };
}

export function createPresets(defs, namespaceId, t) {
  if (!defs.length) {
    return null;
  }

  return [
    {
      name: t('common.labels.default-preset'),
      value: createConfigMapTemplate(namespaceId),
    },
    ...defs.map(({ name, ...value }) => ({
      name,
      value,
    })),
  ];
}
