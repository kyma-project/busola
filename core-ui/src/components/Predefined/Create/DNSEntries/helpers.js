import * as jp from 'jsonpath';

export function dnsEntryToYaml(dnsEntry) {
  return {
    apiVersion: 'dns.gardener.cloud/v1alpha1',
    kind: 'DNSEntry',
    metadata: {
      name: dnsEntry.name,
      namespace: dnsEntry.namespace,
      labels: dnsEntry.labels,
      annotations: dnsEntry.annotations,
    },
    spec: {
      dnsName: dnsEntry.dnsName,
      targets: dnsEntry.targets,
      text: dnsEntry.text,
      ttl: dnsEntry.ttl,
    },
  };
}

export function yamlToDNSEntry(yaml) {
  return {
    name: jp.value(yaml, '$.metadata.name') || '',
    namespace: jp.value(yaml, '$.metadata.namespace') || '',
    labels: jp.value(yaml, '$.metadata.labels') || {},
    annotations: jp.value(yaml, '$.metadata.annotations') || {},
    dnsName: jp.value(yaml, '$.spec.dnsName') || '',
    targets: jp.value(yaml, '$.spec.targets') || [],
    text: jp.value(yaml, '$.spec.text') || [],
    ttl: jp.value(yaml, '$.spec.ttl') || 600,
  };
}

export function createDNSEntryTemplate(namespaceId) {
  return {
    name: '',
    namespace: namespaceId,
    labels: {},
    annotations: {},
    dnsName: '',
    targets: [],
    text: [],
    ttl: 600,
  };
}
