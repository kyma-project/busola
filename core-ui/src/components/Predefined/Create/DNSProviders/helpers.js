import * as jp from 'jsonpath';

export function dnsProviderToYaml(dnsProvider) {
  return {
    apiVersion: 'dns.gardener.cloud/v1alpha1',
    kind: 'DNSProvider',
    metadata: {
      name: dnsProvider.name,
      namespace: dnsProvider.namespace,
      labels: dnsProvider.labels,
    },
    spec: {
      domains: {
        include: dnsProvider.domains?.include?.filter(d => d) || [],
        exclude: dnsProvider.domains?.exclude?.filter(d => d) || [],
      },
      secretRef: dnsProvider.secretRef,
      defaultTTL: dnsProvider.defaultTTL,
    },
  };
}

export function yamlToDNSProvider(yaml) {
  return {
    name: jp.value(yaml, '$.metadata.name') || '',
    namespace: jp.value(yaml, '$.metadata.namespace') || '',
    labels: jp.value(yaml, '$.metadata.labels') || {},
    domains: jp.value(yaml, '$.spec.domains') || '',
    secretRef: jp.value(yaml, '$.spec.secretRef') || {
      name: '',
      namespace: '',
    },
    defaultTTL: jp.value(yaml, '$.spec.defaultTTL') || '',
  };
}

export function createDNSProviderTemplate(namespaceId) {
  return {
    name: '',
    namespace: namespaceId,
    labels: {},
    domains: {
      include: [],
      exclude: [],
    },
    defaultTTL: 600,
    secretRef: { name: '', namespace: namespaceId },
  };
}
