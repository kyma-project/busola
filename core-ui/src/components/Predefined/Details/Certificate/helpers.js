import LuigiClient from '@luigi-project/client';

export const goToIssuer = ({ name, namespace }) =>
  LuigiClient.linkManager()
    .fromContext('cluster')
    .navigate(`namespaces/${namespace}/issuers/details/${name}`);

export const goToSecret = ({ name, namespace }) =>
  LuigiClient.linkManager()
    .fromContext('cluster')
    .navigate(`namespaces/${namespace}/secrets/details/${name}`);
