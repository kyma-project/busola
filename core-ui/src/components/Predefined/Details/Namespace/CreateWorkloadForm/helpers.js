export function formatDeployment(deployment) {
  if (!deployment.labels.app) {
    deployment.labels.app = deployment.name;
  }

  const runtimeDeployment = {
    apiVersion: 'apps/v1',
    kind: 'Deployment',
    metadata: {
      name: deployment.name,
      namespace: deployment.namespace,
      labels: deployment.labels,
    },
    spec: {
      replicas: 1,
      selector: { matchLabels: deployment.labels },
      template: {
        metadata: { labels: deployment.labels },
        spec: {
          containers: [
            {
              name: deployment.name,
              image: deployment.dockerImage,
              resources: {
                requests: deployment.requests,
                limits: deployment.limits,
              },
            },
          ],
        },
      },
    },
  };
  return runtimeDeployment;
}

export function formatService(deployment, deploymentUID) {
  const service = {
    apiVersion: 'v1',
    kind: 'Service',
    metadata: {
      name: deployment.name,
      namespace: deployment.namespace,
      ownerReferences: [
        {
          kind: 'Deployment',
          apiVersion: 'apps/v1',
          name: deployment.name,
          uid: deploymentUID,
        },
      ],
    },
    spec: {
      selector: { app: deployment.name },
      type: 'ClusterIP',
      ports: [
        {
          name: 'http',
          port: deployment.port.port,
          protocol: 'TCP',
          targetPort: deployment.port.targetPort,
        },
      ],
    },
  };
  return service;
}

export const createDeploymentTemplate = namespaceId => ({
  name: '',
  namespace: namespaceId,
  createService: true,
  dockerImage: '',
  labels: {},
  requests: {
    memory: '64Mi',
    cpu: '50m',
  },
  limits: {
    memory: '128Mi',
    cpu: '100m',
  },
  port: {
    name: 'http',
    port: 80,
    protocol: 'TCP',
    targetPort: 8080,
  },
});
