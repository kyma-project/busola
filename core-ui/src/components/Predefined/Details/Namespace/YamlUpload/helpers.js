export const getResourceKindUrl = resource => {
  return `/${resource?.apiVersion === 'v1' ? 'api' : 'apis'}/${
    resource?.apiVersion
  }`;
};

export const getResourceUrl = (resource, namespace) => {
  return `${getResourceKindUrl(resource)}${
    namespace
      ? '/namespaces/' + namespace
      : resource?.metadata?.namespace
      ? '/namespaces/' + resource.metadata.namespace
      : ''
  }/${resource?.kind?.toLowerCase()}s`;
};
