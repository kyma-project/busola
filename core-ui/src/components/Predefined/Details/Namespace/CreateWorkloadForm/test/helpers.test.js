import { formatDeployment, formatService } from '../helpers';

describe('CreateWorkloadForm helpers', () => {
  it('creates uploadable deployment', () => {
    const deployment = {
      name: 'test-name',
      namespace: 'test-namespace',
      createService: false,
      dockerImage: 'test-image',
      labels: { 'test-a': 'test-b' },
      requests: {
        memory: '60Mi',
        cpu: '51m',
      },
      limits: {
        memory: '128Mi',
        cpu: '1000m',
      },
    };

    const formatted = formatDeployment(deployment);

    expect(formatted).toMatchSnapshot();
  });

  it('creates uploadable service', () => {
    const deployment = {
      name: 'test-name',
      namespace: 'test-namespace',
      createService: false,
      dockerImage: 'test-image',
      labels: {},
      port: {
        name: 'http',
        port: 80,
        protocol: 'TCP',
        targetPort: 8080,
      },
    };

    const formatted = formatService(deployment, 'test-id');

    expect(formatted).toMatchSnapshot();
  });
});
