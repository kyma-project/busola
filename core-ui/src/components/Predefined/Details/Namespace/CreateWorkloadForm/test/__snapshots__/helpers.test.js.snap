// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWorkloadForm helpers creates uploadable deployment 1`] = `
Object {
  "apiVersion": "apps/v1",
  "kind": "Deployment",
  "metadata": Object {
    "labels": Object {
      "app": "test-name",
      "test-a": "test-b",
    },
    "name": "test-name",
    "namespace": "test-namespace",
  },
  "spec": Object {
    "replicas": 1,
    "selector": Object {
      "matchLabels": Object {
        "app": "test-name",
        "test-a": "test-b",
      },
    },
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "app": "test-name",
          "test-a": "test-b",
        },
      },
      "spec": Object {
        "containers": Array [
          Object {
            "image": "test-image",
            "name": "test-name",
            "resources": Object {
              "limits": Object {
                "cpu": "1000m",
                "memory": "128Mi",
              },
              "requests": Object {
                "cpu": "51m",
                "memory": "60Mi",
              },
            },
          },
        ],
      },
    },
  },
}
`;

exports[`CreateWorkloadForm helpers creates uploadable service 1`] = `
Object {
  "apiVersion": "v1",
  "kind": "Service",
  "metadata": Object {
    "name": "test-name",
    "namespace": "test-namespace",
  },
  "spec": Object {
    "ports": Array [
      Object {
        "name": "http",
        "port": 80,
        "protocol": "TCP",
        "targetPort": 8080,
      },
    ],
    "selector": Object {
      "app": "test-name",
    },
    "type": "ClusterIP",
  },
}
`;
